core/spec/models/spree/concerns/user_address_book_spec.rb

```diff
@@ -289,82 +371,82 @@ module Spree
     end
 
     context "#persist_order_address" do
-      it 'will save the bill/ship_address reference if it can' do
-        order = create :order
-        user.persist_order_address(order)
-
-        expect( user.bill_address_id ).to eq order.bill_address_id
-        expect( user.ship_address_id ).to eq order.ship_address_id
-        expect( user.bill_address_id ).not_to eq user.ship_address_id
-      end
-
       context "when automatic_default_address preference is at a default of true" do
-        before do
-          stub_spree_preferences(automatic_default_address: true)
-          expect(user).to receive(:save_in_address_book).with(kind_of(Hash), true)
-          expect(user).to receive(:save_in_address_book).with(kind_of(Hash), false)
-        end
+        let(:order) { build :order }
 
-        it "does set the default: true flag" do
-          order = build(:order)
+        it 'will save both bill/ship_address references' do
           user.persist_order_address(order)
+
+          expect( user.bill_address_id ).to eq order.bill_address_id
+          expect( user.ship_address_id ).to eq order.ship_address_id
+          expect( user.bill_address_id ).not_to eq user.ship_address_id
+
+          expect( user.bill_address).to eq order.bill_address
+          expect( user.ship_address).to eq order.ship_address
         end
       end
 
       context "when automatic_default_address preference is false" do
+        let(:order) { build :order }
+
         before do
           stub_spree_preferences(automatic_default_address: false)
-          expect(user).to receive(:save_in_address_book).with(kind_of(Hash), false).twice
-          # and not the optional 2nd argument
         end
 
-        it "does not set the default: true flag" do
-          order = build(:order)
+        it "will save only the default ship address on user as it is the first address at all" do
           user.persist_order_address(order)
+
+          expect( user.bill_address_id ).to eq order.bill_address_id
+          expect( user.ship_address_id ).to eq order.ship_address_id
+
+          expect( user.bill_address).to be_nil
+          expect( user.ship_address).to eq order.ship_address
         end
       end
 
-      context "when address is nil" do
+      context "when one order address is nil" do
```

Comment:
I think `when the order` sounds more correct

Reply:
I wanted to keep the emphasis on the case when one of the two addresses inside the order is nil. I rephrased it for better clarity (or did I?).